CRAYSAT-1399: Scan images using Snyk when building

ekamin-hpe opened the pull request:
* Add code to release.sh that calls updated release library functions,
  and creates separate artifacts for snyk scan archive and excel
  spreadsheet.
* For convenience, move function that renders templated files into a
  separate script.
* Modify Jenkinsfile to include SNYK_TOKEN, and to publish snyk
  artifacts.

Test Description: I created a temporary branch with updated
hpc-shastarelm-release library code venored in, then built the product
stream.

## Summary and Scope

_Summarize what has changed. Explain why this PR is necessary. What is impacted? Is this a new feature, critical bug fix, etc?_

_Is this change backwards incompatible, backwards compatible, or a backwards compatible bugfix?_

## Issues and Related PRs

_List and characterize relationship to Jira/Github issues and other pull requests. Be sure to list dependencies._

* Resolves [CRAYSAT-1399](https://jira-pro.its.hpecorp.net:8443/browse/CRAYSAT-1399)

## Testing

See commit message

## Risks and Mitigations

_Are there known issues with these changes? Any other special considerations?_


## Pull Request Checklist

- [X] Version number(s) incremented, if applicable
- [X] Copyrights updated
- [X] License file intact
- [X] Target branch correct
- [X] CHANGELOG.md updated
- [X] Testing is appropriate and complete, if applicable
- [X] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

jack-stanek-hpe commented on render_templates.sh:
```diff
@@ -0,0 +1,54 @@
+
+# Set to "unstable" for artifacts from main branch, "stable" for artifacts from
+# release branch.
+SAT_REPO_TYPE="unstable"
```
I realize this is just copied out from `release.sh`, but it might be nice while we're looking at this to automate this bit, for example by wildcard globbing the branch against `release/*` or something.

ekamin-hpe replied:
Agreed, this line particular will need special attention if we backport this change to the release branch.

ekamin-hpe replied:
> for example by wildcard globbing the branch against release/* or something.

I think this would end up giving us problems when creating builds based off release branches since this value is used to find the locations of artifacts. We would have to do something like figuring out of we were based off a release branch or an unstable branch. I'm going to see if we've discussed this before because this feels familiar.

haasken-hpe replied:
We have discussed it before, and for the reason you mention (getting it to work correctly on branches created from release branches vs. main), we just decided to hard-code it to "unstable" and update it when creating release branches.

jack-stanek-hpe commented on release.sh:
```diff
@@ -107,6 +80,22 @@ sed -e "s/@RELEASE@/${RELEASE}/g" "${ROOTDIR}/nexus-repositories.yaml" | generat
+    for result in $snyk_results; do
+        snyk-to-html "$result"
+    done
+    cd -
```
This is pretty neat, I did not know that `cd -` has this behavior!

haasken-hpe replied:
I tend to see `pushd` and `popd` used more frequently in scripts. Seems like one advantage would be that if there were any intermediate `cd`s this way would break, but the `pushd` and `popd` approach is more explicit about which directories to record on the stack.

jack-stanek-hpe commented:
I see the Jenkins builds are failing, it looks like that is because the vendored `hpc-shastarelm-release` library needs to be updated for this to work?

ekamin-hpe commented on Jenkinsfile.github:
```diff
@@ -44,6 +44,10 @@ pipeline {
     stages {
         stage('Build release distribution') {
+            environment {
+                // See https://githubmemory.com/repo/jenkinsci/snyk-security-scanner-plugin/issues/107
```
Oops, this was blindly taken from https://github.com/Cray-HPE/csm/blob/main/Jenkinsfile.github#L109

I'll try and find a reference to snyk-security-scanner-plugin issue 107, I don't particularly like linking to this githubmemory.com site here.

ekamin-hpe replied:
I'm just going to remove the comment. I can't imagine what this issue might have been, the release script is using the `snyk` CLI so either you would have to run `snyk auth` or have the environment variable present.

ekamin-hpe commented:
> I see the Jenkins builds are failing, it looks like that is because the vendored `hpc-shastarelm-release` library needs to be updated for this to work?

Correct, I'll have to add that when the hpc-shastarelm-release PR is merged.

haasken-hpe commented on release.sh:
```diff
@@ -38,39 +38,12 @@ set -ex
 requires rsync sed realpath rpm2cpio
 
 BUILDDIR="${1:-"$(realpath -m "$ROOTDIR/dist/${RELEASE}")"}"
+SNYK_SCAN_DIR="${BUILDDIR}/scans"
```
So the snyk scan results are actually packaged into our release distribution file? Is that intentional? Is there any potential for sensitive info to leak here?

I see below they're also being packaged into a separate tar.gz file. Was the intent to remove them from the `BUILDDIR` when done?

ekamin-hpe replied:
I think in either case we would be publishing the information to artifactory, so I am not sure it's any worse to be packaging it directly in the release distribution.

I think this is the same as CSM/SMA are doing: https://github.com/Cray-HPE/csm/blob/main/release.sh#L306-L316

Similarly, I think it was intentional on the part of CSM to have the scans both in the release distribution as well as in a separate tar file.

ekamin-hpe replied:
Actually, the SMA team has chosen to remove the scans from the release distribution: https://github.hpe.com/hpe/hpc-car-sma-product-stream/blob/release/sma-1.7/release.sh#L125-L126

I'd be happy to implement that change here if you think it's better, I'm personally not sure.

ekamin-hpe replied:
I'm going to go ahead with that change, I don't there's any reason to have the information duplicated.

haasken-hpe replied:
I agree, thanks

ekamin-hpe commented:
I rebased this to pull in the license check workflow as well as to squash the current fixup commit. I also updated the vendor library now that the change to hpc-shastarelm-release has been merged. At this point the build should be succeeding.

haasken-hpe reviewed (approved):
Questions about packaging of snyk results into release distribution.

I'd also suggest a bump to `.version`
